pkg/descheduler/descheduler.go

```diff
@@ -77,13 +77,29 @@ func RunDeschedulerStrategies(rs *options.DeschedulerServer, deschedulerPolicy *
 			return
 		}
 
-		nodePodCount := utils.InitializeNodePodCount(nodes)
+		podEvictor := evictions.NewPodEvictor(
+			rs.Client,
+			evictionPolicyGroupVersion,
+			rs.DryRun,
+			rs.MaxNoOfPodsToEvictPerNode,
+			nodes,
+		)
 
-		strategies.RemoveDuplicatePods(rs, deschedulerPolicy.Strategies["RemoveDuplicates"], evictionPolicyGroupVersion, nodes, nodePodCount)
-		strategies.LowNodeUtilization(rs, deschedulerPolicy.Strategies["LowNodeUtilization"], evictionPolicyGroupVersion, nodes, nodePodCount)
-		strategies.RemovePodsViolatingInterPodAntiAffinity(rs, deschedulerPolicy.Strategies["RemovePodsViolatingInterPodAntiAffinity"], evictionPolicyGroupVersion, nodes, nodePodCount)
-		strategies.RemovePodsViolatingNodeAffinity(rs, deschedulerPolicy.Strategies["RemovePodsViolatingNodeAffinity"], evictionPolicyGroupVersion, nodes, nodePodCount)
-		strategies.RemovePodsViolatingNodeTaints(rs, deschedulerPolicy.Strategies["RemovePodsViolatingNodeTaints"], evictionPolicyGroupVersion, nodes, nodePodCount)
+		if s := deschedulerPolicy.Strategies["RemoveDuplicates"]; s.Enabled {
+			strategies.RemoveDuplicatePods(rs.Client, nodes, rs.EvictLocalStoragePods, podEvictor)
+		}
+		if s := deschedulerPolicy.Strategies["LowNodeUtilization"]; s.Enabled {
+			strategies.LowNodeUtilization(rs.Client, s, nodes, rs.EvictLocalStoragePods, podEvictor)
+		}
+		if s := deschedulerPolicy.Strategies["RemovePodsViolatingInterPodAntiAffinity"]; s.Enabled {
+			strategies.RemovePodsViolatingInterPodAntiAffinity(rs.Client, nodes, rs.EvictLocalStoragePods, podEvictor)
+		}
+		if s := deschedulerPolicy.Strategies["RemovePodsViolatingNodeAffinity"]; s.Enabled {
+			strategies.RemovePodsViolatingNodeAffinity(rs.Client, s, nodes, rs.EvictLocalStoragePods, podEvictor)
+		}
+		if s := deschedulerPolicy.Strategies["RemovePodsViolatingNodeTaints"]; s.Enabled {
+			strategies.RemovePodsViolatingNodeTaints(rs.Client, nodes, rs.EvictLocalStoragePods, podEvictor)
+		}
```

Comment:
maybe we could initialize a map of `name:strategyFunction` and put these all in a loop? That would look more like what the scheduler does and could open us up for better user configuration in the future (like being able to specify a custom ordering of the strategies to run). Something like
```
type strategyFunction func(clientset.Interface, []v1*Node, bool, *evictions.PodEvictor)

strategyFuncs := map[string]strategyFunction{
  "RemoveDuplicates": strategies.RemoveDuplicatePods,
  "LowNodeUtilization": strategies.LowNodeUtilization,
  "RemovePodsViolatingInterPodAntiAffinity": strategies.RemovePodsViolatingInterPodAntiAffinity,
  "RemovePodsViolatingNodeAffinity": strategies.RemovePodsViolatingNodeAffinity,
  "RemovePodsViolatingNodeTaints": strategies.RemovePodsViolatingNodeTaints,
}

for name, f := range strategyFuncs {
  if s := deschedulerPolicy.Strategies[name]; s.Enabled {
    f(rs.Client, nodes, rs.EvictLocalStoragePods, podEvictor)
  }
}
```

The map could be declared at the package level so we have a good `DefaultStrategy` map

Reply:
Yeah, I wanted to talk about this in more detail. Right now every strategy has free will about arguments that go in. Should we create an interface for a strategy? Or, use a data type for it as you suggest? Interface might be too restrictive atm. Private `strategyFunction` is a low hanging fruit that can be extended as other strategies goes in. 

tldr; Though, the list of arguments might grow too much. Later, we might implement some initialization/factory part that will get the config and produces a list of enabled strategies so we don't have to check for `s.Enabled` at all. Each strategy might register itself (as is done in the scheduler) as plugings. The factory will just accept rs.EvictLocalStoragePods and rs.Client for now. Each strategy implementing `Run([]*v1.Node, *evictions.PodEvictor)`

Reply:
I agree, there is a lot of discussion that could be had around implementing more of a "framework" for the descheduler strategies. I was just thinking my change could be a quick fix to clean up the chain of if-statements, but with the side benefit of suggesting a starting framework somewhere down the line. I don't think we need to get too much into it for this PR though